sql/sp.cc

```diff
@@ -2493,6 +2565,38 @@ is_package_public_routine(THD *thd,
 }
 
 
+/**
+  Check if a routine has a declaration in the CREATE PACKAGE statement
+  and that its implementation exists in the PACKAGE BODY.
+
+    @param thd      current thd
+    @param db       the database name
+    @param package  the package name
+    @param name     the routine name
+    @param type     the routine type
+    @retval         true,  if both routine declaration and implementation exist
+    @retval         false, either routine declaration or implementation does not exist
+*/
+
+static bool
+is_package_public_routine_with_body(THD *thd,
```

Comment:
I don't think you need that. Currently (before your patch) the search only verifies that a routine exists in the package, and it's missing in the body - it's an error. Here:
```sql
  delimiter $$;
  create package pkg
    function f1() returns int;
  end;
  $$

  create database pkg$$
  create function pkg.f1() returns int return 2$$

  create procedure p1()
  begin
    select pkg.f1();
  end$$
  delimiter ;$$

  call p1();
```
this test does not print '2', it fails with "FUNCTION test.pkg.f1 does not exist"
The name lookup finds f1 in the package pkg and stops there, it does not check the package body. Your implementation should preserve existing behavior, don't check the body, just use is_package_public_routine().